### ISSUE TITLE
Nullable Reference Types: Compiler should not report CS8603 when assignment exists in the right-hand side of a null coalescing operator

### ISSUE THREAD

**Issue by PathogenDavid** [dotnet/roslyn#27009]
**Version Used**: 05/14/18 Nullable Reference Types Preview (`csc` reports `2.8.0.62830 (e595ee27)`) with Visual Studio 15.7.1

A demo solution is available here: [PathogenPlayground/NullCoalescingOperatorCs8603](https://github.com/PathogenPlayground/NullCoalescingOperatorCs8603)

**Steps to Reproduce**:

Compile the following code, where `Foo` is a class:

```csharp
private static Foo? _foo = null;
public static Foo FooProperty
{
    get => _foo ?? (_foo = new Foo()); // warning CS8603: Possible null reference return.
}

public static Foo FooProperty2
{
    get
    {
        Foo? ret = null;
        return ret ?? (ret = new Foo()); // warning CS8603: Possible null reference return.
    }
}
```

**Expected Behavior**:

No warnings are reported.

**Actual Behavior**:

Possible null reference return is reported for the return statement in each property.

----

The readme on [PathogenPlayground/NullCoalescingOperatorCs8603](https://github.com/PathogenPlayground/NullCoalescingOperatorCs8603) contains some additional details.

`FooProperty2` exists to make sure the problem isn't somehow related to https://github.com/dotnet/roslyn/issues/9978

**Comment by PathogenDavid** [comment]
I've also updated [PathogenPlayground/NullCoalescingOperatorCs860](https://github.com/PathogenPlayground/NullCoalescingOperatorCs8603) with non-static variants of the properties since I found out that the behavior differs for `FooProperty4` when the property and field aren't static.

@cston I suspect that the `FooProperty4` issue is more related to #23063 than the null-coalescing problems. Would you prefer I submit a separate issue for it?

**Comment by jcouv** [comment]
[Confirmed](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQDk4BstWFiAJQkiQDsBjCAFQE8AHFAGgBMQBqAHwFgAoAAQBMAAgBiAe3HCA3sIC+fIcIDCfaXwCQ9BAEsAbrAjD+ARgBsYyQH5hAfTCSALMIC8wstiwBuTfwDMxuaWUhLiAAoI4owIMLSOmuq8wskawhoA5hAwfMkpibkFaTpgwgAU9k6ubh44AJQ5hXl2DuLO1RAA7sGltV7yfA2NGvwA7M1OfQD0k8IdUAhkOmTpKgDKABxmAAx+IMJh4sjIOvhGNVjCSKQQlEajAHSDaQq8Gi9yQA===). 
I think this is indeed a bug. Thanks for reporting.

**Comment by jaredpar** [comment]
Issue appears to be fixed now. Assigning to @jaredpar to add a regression test.
